### PR TITLE
Enable eager CUDA graphs for varlen attention

### PR DESCRIPTION
**Pull request by drisspg**
## Summary
not perscriptive just example

AGENT NOTES
Enable eager forward/backward CUDA graphs for packed-document varlen attention by making its metadata shape static.

Varlen kernels already read the actual document boundaries from device-side `cu_seqlens`. The graph only needs a fixed metadata capacity:

- add `training.max_packed_sequences_per_sample`
- pad unused `cu_seqlens` entries with the final packed-token offset, representing zero-length sequences
- use the packed sample length as the static `max_q`/`max_k` upper bound
- reject capacity overflow with an actionable error
- require a capacity when CUDA graphs and varlen attention are enabled
- enable CUDA graphs in the Llama varlen debug config and update existing varlen training configs

Qwen3.5 remains explicitly unsupported because its GatedDeltaNet path requires changing host sequence-offset tuples.

## Why this should not add FA4 work

FA4's dynamic persistent varlen scheduler reads each sequence's actual length from device offsets. Repeated terminal offsets therefore produce zero-length sequences; they do not turn short documents into `seq_len`-length attention operations.

I measured direct FA4 forward on a GB300 with BF16, 8192 total tokens, 16 heads, head dimension 64, causal attention, and metadata capacity 16. `baseline` used exact metadata and the actual maximum document length; `static` used padded metadata and a fixed 8192 upper bound.

| Documents | Actual max length | Static / baseline latency |
|---:|---:|---:|
| 1 | 8192 | 1.00x |
| 2 | 4096 | 0.97x |
| 4 | 2048 | 0.97x |
| 8 | 1024 | 0.97x |
| 16 | 512 | 0.99x |

Padding the offset tensor alone and inflating the host upper bound alone were also neutral in this sweep. This does not reproduce the reported ~3x slowdown; that likely involves a different backward/configuration/backend path rather than zero-length ghost sequences themselves.

## Testing

- `pytest -q tests/unit_tests/test_varlen_metadata.py tests/unit_tests/test_config_manager.py -k "varlen or cuda_graphs"` — 11 passed
- 3-step single-GPU Llama varlen training run:
  - step 1 warmup completed
  - step 2 logged `Recorded CUDA graph`
  - step 3 replay completed
- Updated the existing FSDP + varlen integration definition to exercise CUDA graphs instead of disabling them

The local FA4 checkout has an unrelated backward CuTeDSL compilation failure, so the end-to-end forward/backward graph run used PyTorch's working default varlen backend. The FA4 performance sweep above exercised FA4 directly.

### CODE REVIEW

File: torchtitan/models/common/attention.py [hunk 1]
```diff
@@ -631,29 +631,50 @@ def create_varlen_metadata_for_document(
...
+    provided, unused offsets repeat the final packed-token offset and therefore
+    describe zero-length sequences. This keeps metadata shapes static for CUDA
+    graph replay.
 
     Args:
         positions: Per-token position tensor with shape ``[b, s]``. Positions
             reset to 0 at each document start.
         include_host_offsets: Also materialize cumulative sequence offsets as
             host metadata for kernels that need it.
+        max_sequences_per_sample: Fixed metadata capacity for each batch sample.
```
Comment (drisspg) [inline torchtitan/models/common/attention.py:649]: we can jsut derive this from first input if better

**Comment by drisspg** [comment]
closing for angels -> https://github.com/pytorch/torchtitan/pull/4156

File: torchtitan/models/common/attention.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/models/common/attention.py:649]: would that be safe?
